spec/shared_examples_for_actions_with_github_token.rb

```diff
@@ -0,0 +1,29 @@
+require 'spec_helper'
+
+RSpec.shared_examples 'github_token_initialization' do
+  let(:test_token) { 'Test-GithubToken-1234' }
+
+  describe 'GitHub Token is properly passed to the client' do
+    it 'properly passes the environment variable `GITHUB_TOKEN` to Octokit::Client' do
+      ENV['GITHUB_TOKEN'] = test_token
+      expect(Octokit::Client).to receive(:new).with(access_token: test_token)
+      run_action_without(:github_token)
```

Comment:
This shared example relies on the spec including it (`include_examples 'github_token_initialization'`) defining a `run_action_without` method and a `default_params`. The only way a consumer would find out about this would be by running the tests.

E.g. if we rename `run_action_without` to something else in `setfrozentag_action_spec`, we get:

```
Failures:

  1) Fastlane::Actions::SetfrozentagAction initialize GitHub Token is properly passed to the client properly passes the environment variable `GITHUB_TOKEN` to Octokit::Client
     Failure/Error: run_action_without(:github_token)

     NoMethodError:
       undefined method `run_action_without' for #<RSpec::ExampleGroups::FastlaneActionsSetfrozentagAction::Initialize::GitHubTokenIsProperlyPassedToTheClient:0x0000000109277748>
       Did you mean?  run_action_with
     Shared Example Group: "github_token_initialization" called from ./spec/setfrozentag_action_spec.rb:60
     ...
```

In Swift, we'd do something like constraining this to tests that conform to a certain `protocol`. There's no such thing in vanilla Ruby that I'm aware of.

There might be a way to update `run_described_fastlane_action` to add it the capability to do `_with` and `_without`, but that wouldn't solve the implicit dependency on `default_params` 🤔

At this point, the only thing I can think of is to document the requirement as a comment at the start of the file. What do you think?

Reply:
shared examples can also contain metadata, that you can then pass as additional params when you call `include_examples`. I think that would be a better way to pass `default_params`.

As for `_with` and `_without` variants, I wonder if it's worth relying on those existing in the implementation of `shared_examples` vs just directly using `default_params.except(…)` and `default_params.merge(…)` so that we don't rely on those `_with[out]` methods existing

Reply:
> shared examples can also contain metadata, that you can then pass as additional params when you call include_examples. I think that would be a better way to pass default_params.

Indeed, when I try to use the shared examples passing the `default_params` as metadata, the RSpec blocked me from doing so. The error that I receive is: As we have the `default_params` as a `:let` variable it is not available on an example group (e.g. a `describe` or `context` block). It is only available within individual examples (e.g. `it` blocks). 

In that case, I don't know what is the best alternative here, if is to keep the shared examples relying on the existence of a variable named `default_params` and document this as @mokagio suggested. Or to [redundant use another `:let` variable](https://stackoverflow.com/a/48588994) to pass that metadata as in this example 🤔 

```ruby
  describe 'initialize' do
    include_examples 'github_token_initialization' do
      let(:params) { default_params }
    end
  end
```

> As for _with and _without variants, I wonder if it's worth relying on those existing in the implementation of shared_examples vs just directly using default_params.except(…) and default_params.merge(…) so that we don't rely on those _with[out] methods existing

Yeah, I have that "dilemma" when I was implementing this, I was first defaulting to those approaches of having the `merge` and `except` instead of relying on the `_with` and `_without` functions to exist. However, as I didn't find a good way to pass the params to inside the shared models, I end up following the same pattern that I was in the examples and relying on the existence of those methods as I do not want to duplicate them inside the shared examples.

Reply:
Also, there is this third option to do it, which is cleaner than the other one, but I find it a little bit confusing the way that it used the `send` with the variable name (from a perspective of a person who does not has so much knowledge of ruby 😅) 

``` ruby
shared_examples 'github_token_initialization' do |let_name:|
  let(:params) { send(let_name) }
  ....
end

# spec/test.rb
describe 'initialize' do
 include_examples 'github_token_initialization', let_name: :default_params
end
```

But again, although I feel that this third option is cleaner, I have no strong option about which one is the right one that we should use here.

Reply:
Tbh I'm not a fan of using `send(…)`, both because it's lesser known, but also because it feels like a trick more than a proper official way to do this (the times I've used it was mostly to workaround some limitations rather than to implement something cleanly).

What about using metadata, but workaround the limitation that `let` is only visible in example and not groups… by declaring `default_params` as a function (`def`) rather than a `let`? I know that's what you had in an earlier iteration of the PR before I gave you some feedback that `let` is usually better suited for constant values, and that's still true, but if that is a blocker for us using it in `include_examples`' metadata, then I think it's ok to go back to `def` for that case then. At least I'd personally prefer this over using `send` 🙃

Reply:
Yeah, I find the use of `send` kinda odd too 🙃. 

However,  even if I change `default_params` to a `def` function, I still receive the same error from the RSpec as using it as a `let` variable. 

```ruby
def default_params
{ 
  repository: 'test-repository',
  milestone: '10.1',
  github_token: 'Test-GithubToken-1234' 
}
end

#When I call
describe 'initialize' do
 include_examples 'github_token_initialization', default_params
end

# I receive the same error as the let
Failure/Error: include_examples 'github_token_initialization', default_params
  `default_params` is not available on an example group (e.g. a `describe` or `context` block). It is only available from within individual examples (e.g. `it` blocks) or from constructs that run in the scope of an example (e.g. `before`, `let`, etc).
```

-------------------------

The only way that seems to do the job (although is not a DRY solution), without using the `send`, Is to repeat the declaration of the parameters at the call of the `shared_examples` like this:

``` ruby
describe 'initialize' do
  include_examples 'github_token_initialization',  { repository: 'test-repository', milestone: '10.1', github_token: 'Test-GithubToken-1234' }
end
```

Let me know what you think about this suggestion 😄